scripts/generate_kernel_header.py

```diff
@@ -33,26 +31,33 @@
 namespace uur {
 namespace device_binaries {
     std::map<std::string, std::vector<std::string>> program_kernel_map = {
-% for program, entry_points in kernel_name_dict.items():
-        {"${program}", {
-  % for entry_point in entry_points:
-            "${entry_point}",
-  % endfor
-        }},
-% endfor
+%s
     };
 }
 }
-""")
+"""
+
+PROGRAM_TEMPLATE = """\
+        {"%s", {
+%s
+        }},
+"""
 
+ENTRY_POINT_TEMPLATE = """\
+            "%s",
+"""
 
 def generate_header(output_file, kernel_name_dict):
     """Render the template and write it to the output file."""
     file_name = os.path.basename(output_file)
-    rendered = HEADER_TEMPLATE.render(file_name=file_name,
-                                      kernel_name_dict=kernel_name_dict)
+    device_binaries = ""
+    for program, entry_points in kernel_name_dict.items():
+        content = ""
+        for entry_point in entry_points:
+            content += ENTRY_POINT_TEMPLATE % entry_point
```

Comment:
Is `.format` not preferred over `%` in modern Python?

Reply:
Since this is C++ code the `{}` position markers that `.format()` uses would necessitate escaping all of the braces in the template files. I didn't feel like doing that so chose the easy option as you see here.